model/riscv_pmp_control.sail

```diff
@@ -106,14 +106,42 @@ function pmpCheckRWX(ent, acc) = {
   }
 }
 
-// this needs to be called with the effective current privilege.
-val pmpCheckPerms: (Pmpcfg_ent, AccessType(ext_access_type), Privilege) -> bool
+/* this needs to be called with the effective current privilege */
+val pmpCheckPerms: (Pmpcfg_ent, AccessType(ext_access_type), Privilege) -> bool effect {rreg}
 function pmpCheckPerms(ent, acc, priv) = {
-  match priv {
-    Machine => if   pmpLocked(ent)
-               then pmpCheckRWX(ent, acc)
-               else true,
-    _       => pmpCheckRWX(ent, acc)
+  match mseccfg.MML() {
+    0b1 =>
+      match (ent.R(), ent.W(), ent.L(), ent.X()) {
+        (0b0, 0b1, 0b1, _) =>
+          match acc {
+            Execute(_) => true,
+            Read(_)    => priv == Machine & ent.X() == 0b1,
+            _          => false
+          },
+        (0b0, 0b1, 0b0, _) =>
+          match acc {
+            Read(_)  => true,
+            Write(_) => priv == Machine | ent.X() == 0b1,
+            _        => false
+          },
+        (0b1, 0b1, 0b1, 0b1) =>
+          match acc {
+            Read(_) => true,
+            _       => false
+          },
+        (_, _, _, _) =>
+          if (priv == Machine) == pmpLockBit(ent)
+          then pmpCheckRWX(ent, acc)
```

Comment:
I think this isn't true for LRWX=1010 where X is zero but it is a shared execute only region.

Reply:
It's also not true for LRWX=1011 where the region is readable in M mode.

Reply:
This check is not intended to check LRWX=1010 and LRWX=1011 behaviour. Instead it is checked in the first match statement at line 115 and working as intended.